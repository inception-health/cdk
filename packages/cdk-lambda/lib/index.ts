export { Lambda } from "./lambda";
export type { LambdaProps } from "./lambda";

export { DeadLetterQueue } from "./dlq/queue";
export type { DeadLetterQueueProps } from "./dlq/queue";

export { StandardNetwork } from "./network/network";
export type { StandardNetworkProps } from "./network/network";
