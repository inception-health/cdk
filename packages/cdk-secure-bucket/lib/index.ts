export * from "./s3.secure";
export * from "./lifecycle";
