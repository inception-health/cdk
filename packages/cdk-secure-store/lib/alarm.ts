import {
  Duration,
  aws_cloudwatch,
  aws_cloudwatch_actions,
  aws_dynamodb,
  aws_sns,
} from "aws-cdk-lib";
import { Construct } from "constructs";

import { SecureStore } from "./store";

export interface SecureStoreAlarmsProps {
  /**
   * The SNS topic to send DevOps notifications to. The SNS topic must be
   * created in the same account and region as the DynamoDB table.
   */
  devOpsTopicArn?: string;
  /**
   * Enable System Error Alarm
   *
   * @defaultValue true
   */
  systemErrorAlarm?: boolean;

  /**
   * Enable Throttle Alarm
   *
   * @defaultValue true
   */
  throttleAlarm?: boolean;
}

interface SecureStoreAlarmsCdk {
  table: aws_dynamodb.ITable;
  topic: aws_sns.ITopic | undefined;
  widgets: aws_cloudwatch.IWidget[];
}

const defaultProps: Partial<SecureStoreAlarmsProps> = {
  systemErrorAlarm: true,
  throttleAlarm: true,
};

/**
 * `SecureStoreAlarms` is a construct that creates alarms for monitoring a
 * secure store.
 *
 * This construct is responsible for creating CloudWatch alarms for various
 * metrics related to a secure store.
 *
 * The alarms created by this construct can be used to notify operators when
 * certain conditions are met.
 *
 * @remarks
 * We currently have two alarms:
 * - DynamoDB Table Reads/Writes Throttled: This alarm triggers when the
 *   DynamoDB table is throttled for reads or writes. This alarm is triggered
 *   when the DynamoDB table is throttled for 2% of the time in the last 30
 *   minutes.
 * - DynamoDB Errors > 0: This alarm triggers when there are any errors in the
 *   DynamoDB table. This alarm is triggered when there are any errors in the
 *   DynamoDB table in the last 30 minutes.
 *
 * @example
 * ```
 * // Creating a new instance of SecureStoreAlarms
 * const secureStoreAlarms = new SecureStoreAlarms(stack, 'MySecureStoreAlarms', {
 *   secureStore: mySecureStore,
 * });
 * ```
 *
 * @see
 * https://github.com/cdk-patterns/serverless/blob/main/the-cloudwatch-dashboard/README.md
 */
export class SecureStoreAlarms extends Construct {
  public readonly cdk: SecureStoreAlarmsCdk;

  // !Private Properties
  private readonly props: SecureStoreAlarmsProps;

  constructor(scope: SecureStore, id: string, props: SecureStoreAlarmsProps) {
    super(scope, id);

    // Initialize the properties
    this.props = { ...defaultProps, ...props };
    this.cdk = {} as SecureStoreAlarmsCdk;
    this.cdk.widgets = [];

    this.cdk.table = scope.cdk.table;

    if (this.props.devOpsTopicArn)
      this.cdk.topic = aws_sns.Topic.fromTopicArn(
        this,
        "DevOpsTopic",
        this.props.devOpsTopicArn,
      );

    // ==================================
    // System Errors
    // ==================================
    if (this.props.systemErrorAlarm) this.setupSystemError();

    // ==================================
    // Read/Write Throttles
    // ==================================
    if (this.props.throttleAlarm) this.setupThrottle();

    // ==================================
    // Cloudwatch Widgets
    // ==================================
    this.setupCloudwatchWidgets();
  }

  // !Private Methods
  private buildGraphWidget(
    widgetName: string,
    metrics: aws_cloudwatch.IMetric[],
    stacked = false,
  ): aws_cloudwatch.GraphWidget {
    const widget = new aws_cloudwatch.GraphWidget({
      title: widgetName,
      left: metrics,
      stacked: stacked,
      width: 8,
    });
    this.cdk.widgets.push(widget);
    return widget;
  }

  // !Private Setup
  private setupSystemError() {
    // ----------------------------------
    // Metric
    // ----------------------------------
    const metric = this.cdk.table.metricSystemErrorsForOperations({
      operations: [
        aws_dynamodb.Operation.GET_ITEM,
        aws_dynamodb.Operation.PUT_ITEM,
        aws_dynamodb.Operation.DELETE_ITEM,
        aws_dynamodb.Operation.UPDATE_ITEM,
        aws_dynamodb.Operation.TRANSACT_WRITE_ITEMS,
        aws_dynamodb.Operation.TRANSACT_GET_ITEMS,
        aws_dynamodb.Operation.BATCH_GET_ITEM,
        aws_dynamodb.Operation.BATCH_WRITE_ITEM,
        aws_dynamodb.Operation.QUERY,
        aws_dynamodb.Operation.SCAN,
      ],
      period: Duration.minutes(5),
    });

    // ----------------------------------
    // Alarm
    // ----------------------------------
    const alarm = new aws_cloudwatch.Alarm(this, "alarm-errors", {
      alarmName: `DynamoDB ${this.cdk.table.tableName} System Errors`,
      // Metric
      metric: metric,
      // Configuration
      threshold: 0,
      evaluationPeriods: 6,
      datapointsToAlarm: 1,
      treatMissingData: aws_cloudwatch.TreatMissingData.NOT_BREACHING,
    });

    // ----------------------------------
    // Actions
    // ----------------------------------
    // If a topic is defined in the CDK stack, an SNS action is added to the
    // alarm. This means that when the alarm state changes, a message will be
    // published to the SNS topic.
    if (this.cdk.topic)
      alarm.addAlarmAction(
        new aws_cloudwatch_actions.SnsAction(this.cdk.topic),
      );
  }

  private setupThrottle() {
    // ----------------------------------
    // Metric
    // ----------------------------------
    // The purpose of this metric is to monitor the throttling events of
    // DynamoDB. It combines the `ReadThrottleEvents` and `WriteThrottleEvents` into
    // a single metric. The statistic used is `'sum'`, which means it will add up
    // all the throttle events in the specified period. The period is set to 5
    // minutes, meaning the metric data points are emitted every 5 minutes.
    const metric = new aws_cloudwatch.MathExpression({
      expression: "m1 + m2",
      label: `DynamoDB ${this.cdk.table.tableName} Reads/Writes Throttled`,
      usingMetrics: {
        m1: this.cdk.table.metric("ReadThrottleEvents", { statistic: "sum" }),
        m2: this.cdk.table.metric("WriteThrottleEvents", { statistic: "sum" }),
      },
      period: Duration.minutes(5),
    });

    // ----------------------------------
    // Alarm
    // ----------------------------------
    // The purpose of this alarm is to trigger when there are any DynamoDB
    // throttling events. The alarm is configured to trigger when the metric
    // value is greater than 0. The evaluationPeriods is set to 6, meaning the
    // alarm will look at the last 6 periods (or 30 minutes) of data. The
    // datapointsToAlarm is set to 1, meaning the alarm will trigger if any
    // single data point in the evaluation period breaches the threshold. The
    // treatMissingData is set to NOT_BREACHING, meaning if there are any
    // periods of missing data, they will not count as breaching the threshold.
    const alarm = new aws_cloudwatch.Alarm(this, "alarm-throttle", {
      alarmName: `DynamoDB ${this.cdk.table.tableName} Throttled`,
      // Metric
      metric: metric,
      // Configuration
      threshold: 0,
      evaluationPeriods: 6,
      datapointsToAlarm: 1,
      treatMissingData: aws_cloudwatch.TreatMissingData.NOT_BREACHING,
    });

    // ----------------------------------
    // Actions
    // ----------------------------------
    // If a topic is defined in the CDK stack, an SNS action is added to the
    // alarm. This means that when the alarm state changes, a message will be
    // published to the SNS topic.
    if (this.cdk.topic)
      alarm.addAlarmAction(
        new aws_cloudwatch_actions.SnsAction(this.cdk.topic),
      );
  }

  private setupCloudwatchWidgets() {
    // ----------------------------------
    // DynamoDB Latency
    // ----------------------------------
    this.buildGraphWidget(
      `DynamoDB ${this.cdk.table.tableName} Latency`,
      ["GetItem", "UpdateItem", "PutItem", "DeleteItem", "Query"].map((op) => {
        return this.cdk.table.metricSuccessfulRequestLatency({
          dimensionsMap: {
            TableName: this.cdk.table.tableName,
            Operation: op,
          },
        });
      }),
      true,
    );

    // ----------------------------------
    // DynamoDB Consumed Read/Write Units
    // ----------------------------------
    this.buildGraphWidget(
      `DynamoDB ${this.cdk.table.tableName} Consumed Read/Write Units`,
      [
        this.cdk.table.metricConsumedReadCapacityUnits(),
        this.cdk.table.metricConsumedWriteCapacityUnits(),
      ],
      true,
    );

    // ----------------------------------
    // DynamoDB Throttles
    // ----------------------------------
    this.buildGraphWidget(
      `DynamoDB ${this.cdk.table.tableName} Throttles`,
      [
        this.cdk.table.metric("ReadThrottleEvents", { statistic: "sum" }),
        this.cdk.table.metric("WriteThrottleEvents", { statistic: "sum" }),
      ],
      true,
    );
  }
}
