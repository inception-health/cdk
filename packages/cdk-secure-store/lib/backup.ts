import { aws_backup, aws_dynamodb, aws_kms } from "aws-cdk-lib";
import { Construct } from "constructs";

export interface SecureStoreBackupProps {
  backupPlanId?: string;
  encryptionKey: aws_kms.IKey;
  table: aws_dynamodb.ITableV2;
  tableName: string;
}

export interface SecureStoreBackupResources {
  backupPlan: aws_backup.IBackupPlan | undefined;
  backupVault: aws_backup.BackupVault | undefined;
}

export class SecureStoreBackup extends Construct {
  public readonly cdk: SecureStoreBackupResources;

  private readonly props: SecureStoreBackupProps;

  constructor(scope: Construct, id: string, props: SecureStoreBackupProps) {
    super(scope, id);

    // Initialize the properties
    this.props = props;
    this.cdk = {} as SecureStoreBackupResources;

    // Create the backup resources
    if (this.props.backupPlanId) {
      this.cdk.backupPlan = aws_backup.BackupPlan.fromBackupPlanId(
        this,
        "backupPlan",
        this.props.backupPlanId,
      );
    } else {
      // Create a backup vault
      this.cdk.backupVault = new aws_backup.BackupVault(this, "backupVault", {
        backupVaultName: `secure-store-${this.props.tableName}-backup-vault`,
        encryptionKey: this.props.encryptionKey,
        blockRecoveryPointDeletion: true,
        // TODO: Review if we want to make lock configuration by default.
        // lockConfiguration
      });

      // Create the backup plan
      this.cdk.backupPlan =
        aws_backup.BackupPlan.dailyWeeklyMonthly7YearRetention(
          this,
          `secure-store-${this.props.tableName}-backup-plan`,
          this.cdk.backupVault,
        );

      (this.cdk.backupPlan as aws_backup.BackupPlan).addSelection(
        `secure-store-${this.props.tableName}-backup-plan-selection`,
        {
          resources: [
            aws_backup.BackupResource.fromDynamoDbTable(this.props.table),
          ],
        },
      );
    }
  }
}
