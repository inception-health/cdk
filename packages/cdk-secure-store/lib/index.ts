export type { SecureStoreProps } from "./store";
export { SecureStore } from "./store";
