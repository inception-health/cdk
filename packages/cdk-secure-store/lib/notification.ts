import {
  aws_events,
  aws_events_targets,
  aws_kms,
  aws_logs,
  aws_sns,
} from "aws-cdk-lib";
import { Construct } from "constructs";

export interface SecureStoreNotificationProps {
  /**
   * The SNS Topic to send notifications to.
   */
  destination: string;
  /**
   * Enable CloudWatch Logs for the EventBridge rule.
   *
   * @defaultValue false
   */
  enableLogs?: boolean;
  /**
   * Enable or disable the notification.
   *
   * @defaultValue true
   */
  enabled?: boolean;
  /**
   * The encryption key to use for encrypting the sns topic and the logs
   *
   * @remarks
   * The encryption key must be created in the same account and region as the
   * DynamoDB table. It also must have the following policy attached to it:
   * ```
   * {
   *   "Sid": "Allow access to CloudWatch Log",
   *   "Effect": "Allow",
   *   "Principal": {
   *       "Service": "logs.REGION.amazonaws.com"
   *   },
   *   "Action": [
   *       "kms:Encrypt",
   *       "kms:Decrypt",
   *       "kms:ReEncrypt*",
   *       "kms:GenerateDataKey*",
   *       "kms:DescribeKey"
   *   ],
   *   "Resource": "*",
   *   "Condition": {
   *       "ArnLike": {
   *           "kms:EncryptionContext:aws:logs:arn": "arn:aws:logs:REGION:ACCOUNT_ID:*"
   *       }
   *   }
   * }
   *```
   */
  encryptionKey: aws_kms.IKey;
  /**
   * The secure store table name
   */
  sourceTable: string;
}

const defaultProps: Partial<SecureStoreNotificationProps> = {
  enabled: true,
  enableLogs: true,
};

interface SecureStoreNotificationResources {
  bus: aws_events.IEventBus;
  logs: aws_logs.LogGroup;
  rule: aws_events.Rule;
  topic: aws_sns.ITopic;
}

/**
 * `Notification` is a construct that creates a notification for specific
 *  DynamoDB events. The notification is sent to the provided SNS topic.
 *
 * At the moment, the notification is sent for the following events:
 * - CreateTable
 * - DeleteTable
 * - UpdateTable
 * - DeleteBackup
 * - RestoreTableFromBackup
 * - RestoreTableToPointInTime
 *
 * @remarks
 * For the full list of events that trigger a notification, see the [DynamoDB
 * documentation](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/logging-using-cloudtrail.html).
 */
export class SecureStoreNotification extends Construct {
  public readonly cdk: SecureStoreNotificationResources;

  // !Private Properties
  private readonly props: SecureStoreNotificationProps;

  // !Constructor
  constructor(
    scope: Construct,
    id: string,
    props: SecureStoreNotificationProps,
  ) {
    super(scope, id);

    // Initialize the properties
    this.props = { ...defaultProps, ...props };
    this.cdk = {} as SecureStoreNotificationResources;

    // Create Resources
    this.setupTopic();

    this.setupBus();

    this.setupRule();

    this.setupLogs();
  }

  // !Private Methods
  private setupTopic() {
    this.cdk.topic = aws_sns.Topic.fromTopicArn(
      this,
      "topic",
      this.props.destination,
    );
  }

  private setupBus() {
    this.cdk.bus = aws_events.EventBus.fromEventBusName(
      this,
      "eventBus",
      "default",
    );
  }

  private setupRule() {
    this.cdk.rule = new aws_events.Rule(this, "rule", {
      ruleName: `${this.props.sourceTable}-dynamodb--monitoring`,
      description: `Send system notifications related to the ${this.props.sourceTable} table.`,
      enabled: this.props.enabled,
      eventBus: this.cdk.bus,
      eventPattern: {
        source: ["aws.dynamodb"],
        detailType: ["AWS API Call via CloudTrail"],
        detail: {
          eventSource: ["dynamodb.amazonaws.com"],
          eventName: [
            "CreateTable",
            "DeleteTable",
            "UpdateTable",
            "DeleteBackup",
            "RestoreTableFromBackup",
            "RestoreTableToPointInTime",
          ],
          requestParameters: {
            tableName: [this.props.sourceTable],
          },
        },
      },
    });

    // Add the target to the rule
    this.cdk.rule.addTarget(new aws_events_targets.SnsTopic(this.cdk.topic));
  }

  private setupLogs() {
    if (!this.props.enableLogs) return;

    // New Cloudwatch Log Group to deliver EventBridge Events to.
    this.cdk.logs = new aws_logs.LogGroup(this, "logs", {
      logGroupName: `/inception/devops/${this.props.sourceTable}-dynamodb--monitoring`,
      retention: aws_logs.RetentionDays.TWO_WEEKS,
      encryptionKey: this.props.encryptionKey,
    });

    // Define rule that will deliver EventBridge events to a Cloudwatch Log Group
    this.cdk.rule.addTarget(
      new aws_events_targets.CloudWatchLogGroup(this.cdk.logs),
    );
  }
}
