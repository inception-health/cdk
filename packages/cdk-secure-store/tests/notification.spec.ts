import { Stack, aws_kms } from "aws-cdk-lib";
import { Match, Template } from "aws-cdk-lib/assertions";

import { SecureStoreNotification } from "../lib/notification";

/**
 * In this test, we are testing the Notification construct.
 *
 * - It checks if a topic is created from the destination ARN.
 * - It checks if a rule is created with the correct properties.
 * - It checks if the topic is added as a target to the rule.
 */
describe("Notification", () => {
  let stack: Stack;
  let destination: string;
  let encryptionKey: aws_kms.IKey;

  beforeEach(() => {
    stack = new Stack();
    destination = "arn:aws:sns:us-east-1:123456789012:topic";
    encryptionKey = aws_kms.Key.fromKeyArn(
      stack,
      "key",
      "arn:aws:kms:us-east-1:123456789012:key/12345678-1234-1234-1234-123456789012",
    );
  });
  it("should match the snapshot", () => {
    // When
    new SecureStoreNotification(stack, "notification", {
      destination,
      sourceTable: "MyTable",
      encryptionKey: encryptionKey,
    });

    // Then
    expect(Template.fromStack(stack)).toMatchSnapshot();
  });

  it("should assign the destination ARN", () => {
    // When
    const notification = new SecureStoreNotification(stack, "notification", {
      destination,
      sourceTable: "MyTable",
      encryptionKey: encryptionKey,
    });

    // Then
    expect(notification.cdk.topic.topicArn).toEqual(destination);
  });

  it("should create a rule with the correct properties", () => {
    // When
    new SecureStoreNotification(stack, "notification", {
      destination,
      sourceTable: "MyTable",
      encryptionKey: encryptionKey,
    });

    // Then
    const template = Template.fromStack(stack);
    template.allResourcesProperties("AWS::Events::Rule", {
      EventPattern: {
        detail: {
          eventName: [
            "CreateTable",
            "DeleteTable",
            "UpdateTable",
            "DeleteBackup",
            "RestoreTableFromBackup",
            "RestoreTableToPointInTime",
          ],
          eventSource: ["dynamodb.amazonaws.com"],
          requestParameters: {
            tableName: ["MyTable"],
          },
        },
        "detail-type": ["AWS API Call via CloudTrail"],
        source: ["aws.dynamodb"],
      },
    });
  });

  it("should add the topic as a target to the rule", () => {
    // When
    new SecureStoreNotification(stack, "notification", {
      destination,
      sourceTable: "MyTable",
      encryptionKey: encryptionKey,
    });

    // Then
    const template = Template.fromStack(stack);
    template.hasResourceProperties("AWS::Events::Rule", {
      Targets: [
        {
          Arn: destination,
          Id: "Target0",
        },
        {
          Arn: Match.anyValue(),
          Id: "Target1",
        },
      ],
    });
  });

  it("should not have logging as target if disabled", () => {
    // When
    new SecureStoreNotification(stack, "notification", {
      destination,
      sourceTable: "MyTable",
      encryptionKey: encryptionKey,
      // WHEN
      enableLogs: false,
    });

    // Then
    const template = Template.fromStack(stack);
    template.hasResourceProperties("AWS::Events::Rule", {
      Targets: [
        {
          Arn: destination,
          Id: "Target0",
        },
      ],
    });
  });

  it("should have logging using the encryption key", () => {
    // When
    new SecureStoreNotification(stack, "notification", {
      destination,
      sourceTable: "MyTable",
      // WHEN
      encryptionKey: encryptionKey,
    });

    // Then
    const template = Template.fromStack(stack);
    template.hasResourceProperties("AWS::Logs::LogGroup", {
      KmsKeyId: encryptionKey.keyArn,
    });
  });
});
