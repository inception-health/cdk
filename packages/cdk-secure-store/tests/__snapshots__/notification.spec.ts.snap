// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification should match the snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "AWS679f53fac002430cb0da5b7982bd22872D164C4C": {
      "DependsOn": [
        "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "ed6cd104ff5f101d06dae8cb2b87cc6e6d69b9a22055b467ea6cae10ff023023.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "EventsLogGroupPolicynotificationruleDD8D8A55633545E3": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "EventsLogGroupPolicynotificationruleDD8D8A55CustomResourcePolicyA5BA260C",
      ],
      "Properties": {
        "Create": {
          "Fn::Join": [
            "",
            [
              "{"service":"CloudWatchLogs","action":"putResourcePolicy","parameters":{"policyName":"EventsLogGroupPolicynotificationruleDD8D8A55","policyDocument":"{\\"Statement\\":[{\\"Action\\":[\\"logs:PutLogEvents\\",\\"logs:CreateLogStream\\"],\\"Effect\\":\\"Allow\\",\\"Principal\\":{\\"Service\\":\\"events.amazonaws.com\\"},\\"Resource\\":\\"",
              {
                "Fn::GetAtt": [
                  "notificationlogsFD6D4DEC",
                  "Arn",
                ],
              },
              "\\"}],\\"Version\\":\\"2012-10-17\\"}"},"physicalResourceId":{"id":"EventsLogGroupPolicynotificationruleDD8D8A55"}}",
            ],
          ],
        },
        "Delete": "{"service":"CloudWatchLogs","action":"deleteResourcePolicy","parameters":{"policyName":"EventsLogGroupPolicynotificationruleDD8D8A55"},"ignoreErrorCodesMatching":"ResourceNotFoundException"}",
        "InstallLatestAwsSdk": true,
        "ServiceToken": {
          "Fn::GetAtt": [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
        "Update": {
          "Fn::Join": [
            "",
            [
              "{"service":"CloudWatchLogs","action":"putResourcePolicy","parameters":{"policyName":"EventsLogGroupPolicynotificationruleDD8D8A55","policyDocument":"{\\"Statement\\":[{\\"Action\\":[\\"logs:PutLogEvents\\",\\"logs:CreateLogStream\\"],\\"Effect\\":\\"Allow\\",\\"Principal\\":{\\"Service\\":\\"events.amazonaws.com\\"},\\"Resource\\":\\"",
              {
                "Fn::GetAtt": [
                  "notificationlogsFD6D4DEC",
                  "Arn",
                ],
              },
              "\\"}],\\"Version\\":\\"2012-10-17\\"}"},"physicalResourceId":{"id":"EventsLogGroupPolicynotificationruleDD8D8A55"}}",
            ],
          ],
        },
      },
      "Type": "Custom::CloudwatchLogResourcePolicy",
      "UpdateReplacePolicy": "Delete",
    },
    "EventsLogGroupPolicynotificationruleDD8D8A55CustomResourcePolicyA5BA260C": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "logs:PutResourcePolicy",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "logs:DeleteResourcePolicy",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "EventsLogGroupPolicynotificationruleDD8D8A55CustomResourcePolicyA5BA260C",
        "Roles": [
          {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "notificationlogsFD6D4DEC": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "KmsKeyId": "arn:aws:kms:us-east-1:123456789012:key/12345678-1234-1234-1234-123456789012",
        "LogGroupName": "/inception/devops/MyTable-dynamodb--monitoring",
        "RetentionInDays": 14,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "notificationruleC389A2D9": {
      "Properties": {
        "Description": "Send system notifications related to the MyTable table.",
        "EventBusName": "default",
        "EventPattern": {
          "detail": {
            "eventName": [
              "CreateTable",
              "DeleteTable",
              "UpdateTable",
              "DeleteBackup",
              "RestoreTableFromBackup",
              "RestoreTableToPointInTime",
            ],
            "eventSource": [
              "dynamodb.amazonaws.com",
            ],
            "requestParameters": {
              "tableName": [
                "MyTable",
              ],
            },
          },
          "detail-type": [
            "AWS API Call via CloudTrail",
          ],
          "source": [
            "aws.dynamodb",
          ],
        },
        "Name": "MyTable-dynamodb--monitoring",
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": "arn:aws:sns:us-east-1:123456789012:topic",
            "Id": "Target0",
          },
          {
            "Arn": {
              "Fn::Join": [
                "",
                [
                  "arn:",
                  {
                    "Ref": "AWS::Partition",
                  },
                  ":logs:",
                  {
                    "Ref": "AWS::Region",
                  },
                  ":",
                  {
                    "Ref": "AWS::AccountId",
                  },
                  ":log-group:",
                  {
                    "Ref": "notificationlogsFD6D4DEC",
                  },
                ],
              ],
            },
            "Id": "Target1",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
